source/common/local_reply/local_reply.cc

```diff
@@ -0,0 +1,161 @@
+
+  BodyFormatter(const envoy::config::core::v3::SubstitutionFormatString& config) {
+    formatter_ = SubstitutionFormatStringUtils::fromProtoConfig(config);
+    if (config.format_case() ==
+        envoy::config::core::v3::SubstitutionFormatString::FormatCase::kJsonFormat) {
+      content_type_ = Http::Headers::get().ContentTypeValues.Json;
+    }
+  }
+
+  void format(const Http::RequestHeaderMap& request_headers,
+              const Http::ResponseHeaderMap& response_headers,
+              const Http::ResponseTrailerMap& response_trailers,
+              const StreamInfo::StreamInfo& stream_info, std::string& body,
+              absl::string_view& content_type) const {
+    body =
+        formatter_->format(request_headers, response_headers, response_trailers, stream_info, body);
+    content_type = content_type_;
+  }
+
+private:
+  AccessLog::FormatterPtr formatter_;
+  absl::string_view content_type_{Http::Headers::get().ContentTypeValues.Text};
+};
+
+using BodyFormatterPtr = std::unique_ptr<BodyFormatter>;
+
+class ResponseMapper {
+public:
+  ResponseMapper(
+      const envoy::extensions::filters::network::http_connection_manager::v3::ResponseMapper&
+          config,
+      Server::Configuration::FactoryContext& context) {
+    filter_ = AccessLog::FilterFactory::fromProto(
+        config.filter(), context.runtime(), context.random(), context.messageValidationVisitor());
+
+    if (config.has_status_code()) {
+      status_code_ = static_cast<Http::Code>(config.status_code().value());
+    }
+    if (config.has_body()) {
+      body_ = Config::DataSource::read(config.body(), true, context.api());
+    }
+
+    if (config.has_body_format()) {
+      body_formatter_ = std::make_unique<BodyFormatter>(config.body_format());
+    }
+  }
+
+  bool matchAndRewrite(const Http::RequestHeaderMap& request_headers,
+                       Http::ResponseHeaderMap& response_headers,
+                       const Http::ResponseTrailerMap& response_trailers,
+                       StreamInfo::StreamInfoImpl& stream_info, Http::Code& code, std::string& body,
+                       BodyFormatter*& final_formatter) const {
+    // If not matched, just bail out.
+    if (!filter_->evaluate(stream_info, request_headers, response_headers, response_trailers)) {
+      return false;
+    }
+
+    if (body_.has_value()) {
+      body = body_.value();
+    }
+
+    if (status_code_.has_value() && code != status_code_.value()) {
+      code = status_code_.value();
+      response_headers.setReferenceStatus(std::to_string(enumToInt(code)));
+      stream_info.response_code_ = static_cast<uint32_t>(code);
+    }
+
+    if (body_formatter_) {
+      final_formatter = body_formatter_.get();
+    }
+    return true;
+  }
+
+private:
+  AccessLog::FilterPtr filter_;
+  absl::optional<Http::Code> status_code_;
+  absl::optional<std::string> body_;
+  BodyFormatterPtr body_formatter_;
```

Comment:
nit: perfer const and moving into initializer list if possible, same elsewhere.

Reply:
Done